### ISSUE TITLE
#2752 regrettably introduces a serious bug

### ISSUE THREAD

**Issue by pond** [rspec/rspec-rails#2773]
## What Ruby, Rails and RSpec versions are you using?

Ruby version: 3.3.3
Rails version: 7.1
RSpec version: 6.1.3 (vs working, 6.1.2)

## Observed behaviour

The reset behaviour occurs _in between_ the suite's `config.around :each` (e.g. in `spec_helper.rb`) and any `context`/`describe` block's `before :each ...` (or more generally, before examples then run). This breaks any tests which had current attributes set up in the config, which tests then expect later. This bug is the reason for https://github.com/ErwinM/acts_as_tenant/issues/338.

## Expected behaviour

We strongly suggest that the reset should happen either _after_ all other "each"-style example hooks have concluded, or _before_ any "each"-style example hooks start. It should never happen invisibly _in between_ "each"-style hooks.

## Can you provide an example reproduction?

Yes. A tiny stripped down almost-Rails application with `README.md` containing additional information and a replication test case is included.

[rspecbug.tar.gz](https://github.com/user-attachments/files/16061697/rspecbug.tar.gz)

**Comment by pirj** [comment]
Have you checked if the solution to clean only in after_teardown works?

**Comment by pond** [comment]
@pirj I hadn't, so I just hacked that into the "live" gem via `bundle open`:

```ruby
# ...
      if ::Rails::VERSION::MAJOR >= 7
        include RSpec::Rails::TaggedLoggingAdapter
        def after_teardown
          super
          ActiveSupport::CurrentAttributes.reset_all
        end
        include ActiveSupport::ExecutionContext::TestHelper
      end
# ...
```

(...which is ugly as sin and I'm not suggesting that code formatting style - comments would help, too!) and then:

```
$ bundle exec gem list rspec-rails

*** LOCAL GEMS ***

rspec-rails (6.1.3)

$ bundle exec rspec -f d                   

Reset of ActiveSupport::CurrentAttributes
==> RSpec: config.around :each
++> Class: Setting 'Current.canary' to "spec_helper.rb \"config.around\""
--> Tests: before :each
++> Class: 'Current' WAS RESET
  works

Finished in 0.00118 seconds (files took 0.87574 seconds to load)
1 example, 0 failures
```

...so that looks pretty successful, but do note the caveats that this doesn't really fix the problem, it just pushes it into a more unlikely edge case.

**Comment by pirj** [comment]
Will it push the problem even further if we use [append_after](https://rubydoc.info/gems/rspec-core/RSpec%2FCore%2FHooks:append_after)?

It’s not the first time we have hook order issues.

**Comment by pond** [comment]
@pirj

Sorry for delayed response - timezones; i'm in New Zealand.

Callback ordering can indeed be a royal pain! You need to put something on the tail of the **around**-each callback chain - since around-each is the "outermost" (according to [this reference](https://rspec.info/features/3-13/rspec-core/hooks/around-hooks/)). In pseudocode, what we need to achieve is akin to:

```
append_to_tail_around :each do | example |
  example.run()
  ActiveSupport::CurrentAttributes.reset_all
end
```

RSpec Core should never be depending on anything to do with Rails' current attributes, and any "user" code that tries to append after RSpec Core & RSpec Rails's own callback chains is just asking for trouble anyway. Going to the end of the after-each callback chain is probably not sufficient because the around-each hooks run after those, which is why I think you need to append to the tail of around-each and _that_ should be sufficient. You need to be the "outermost Russian Doll". Beware, though, I'm very much **not** an expert in the implementations of the RSpec family and, depending on how its internal callback chains actually map to what the API exposes as around-vs-after-hooks, `append_after` might be fine.

**Comment by pond** [comment]
NB - If you wish I could try to work up a PR for this. The thing that I hesitate over is testing - I've no idea how to "prove" something breaks before, and works after this fix, given I'd need to do a config-level before-each then an example level before-each and can't figure out how to do emulate that within RSpec Rails's own test suite.

**Comment by pirj** [comment]
Right. So it’s a matter of the right loading of `RSpec.config` blocks that define `config.around` hooks.
But we `config.include` [specific example groups](https://github.com/rspec/rspec-rails/blob/8c17b4e5020a4d264e8a79e294c58b5c1ef2b005/lib/rspec/rails/configuration.rb#L56), and some of them define hooks](https://github.com/rspec/rspec-rails/blob/8c17b4e5020a4d264e8a79e294c58b5c1ef2b005/lib/rspec/rails/example/system_example_group.rb#L173).
An around defined in such way would run before an around defined directly with config.around [like it is in acts_as_tenant](https://github.com/ErwinM/acts_as_tenant/issues/338), so this won’t help.

You can check the ‘features’ directory, there are some examples on how to define config-level hooks.

**Comment by JonRowe** [comment]
A couple of things here, firstly `rspec-rails` is a way to bring test helpers and behaviour from Rails across into RSpec, this means that we try to adopt the same behaviours as testing apps in Rails own test helpers [which are minitest based], so this behaviour is an unfortunate side effect rather than serious bug in my opinion.

We implement minitest's life cycle like this:

```
    module MinitestLifecycleAdapter
      extend ActiveSupport::Concern

      included do |group|
        group.before { after_setup }
        group.after  { before_teardown }

        group.around do |example|
          before_setup
          example.run
          after_teardown
        end
```

This is because of our own precedence around hooks, `around` happen first because they can skip the example and don't have access to the instance, and `before` / `after` happen in the context of the hooks. See the documentation for `around` for warnings: https://rspec.info/features/3-13/rspec-core/hooks/around-hooks/

> WARNING: around hooks do not share state with the example the way before and after hooks do. This means that you cannot share instance variables between around hooks and examples.

> WARNING: Mock frameworks are set up and torn down within the context of running the example. You cannot interact with them directly in around hooks.

> WARNING: around hooks will execute before any before hooks, and after any after hooks regardless of the context they were defined in.

So the problem is actually using an `around` hook to set state, really it should be `before`/ `after` to respect both our and Rails semantics...

**Comment by pond** [comment]
@JonRowe I'm sorry but that's just not a good answer.

You're talking about RSpec Rails invisibly resetting some state *in between* hooks and doing this *before* a test runs. If you at least only did it after, that might help.

I really don't hope this turns into one of those depressing intransigent arguments about design patterns or ethos when what we have here is a bug that's breaking people's tests, is trivially easy to fix, and you have someone offering to do that work for you.

I would hope you'd jump at the chance.

**Comment by pond** [comment]
It's really a question of whether or not we obey semver. The simple unfortunate fact is: a patch version broke working client code that's not doing anything the docs prohibit.

So, we can introduce a breaking change in a patch version and say "forget semver" - which I'd rather strongly object to, NPM shows us the hellscape that we enter if we do that! - or we can come up with a way to resolve the breaking changes. It won't just be our test suite or anyone using ActsAsTenant having issues here and it took us a long time to figure out what was wrong since, of course, you absolutely do not suspect going from 6.1.2 to 6.1.3 of just about anything to have caused the issue, and the manifestation was so bizarre - one minute in our test suite something is set, the next minute it's just gone, and this was invisible to our code.

Very hard to debug.

The short-term fix here would honestly be to keep things simple - *remove this change* and don't reset current attributes. Then the community can discuss how it might want to approach that, either by changing docs, version number strategies, etc. etc. and we can move forward *without RubyGems holding a landmine that people might step on*.

In the mean time I can produce a PR, but Jon has been pretty clear that he wants things to just call existing Rails code so I'm loathe to spend time on this unless he's open to the idea of seeing what the code looks like.  The gem's his baby after all.

**Comment by pirj** [comment]
> just call existing Rails code

"just" is an exaggeration. If you glance over [adapters](https://github.com/rspec/rspec-rails/blob/8c17b4e5020a4d264e8a79e294c58b5c1ef2b005/lib/rspec/rails/adapters.rb#L6) or [fixtures](https://github.com/rspec/rspec-rails/blob/8c17b4e5020a4d264e8a79e294c58b5c1ef2b005/lib/rspec/rails/fixture_support.rb#L1), you'll see it's not always that simple.

So please proceed with the PR with no doubts.

**Comment by pond** [comment]
> So please proceed with the PR with no doubts.

OK, no worries! I'll get onto it (and do the best I can to keep within the spirit of Jon's concerns).

**Comment by pond** [comment]
@JonRowe / @pirj - I've implemented https://github.com/rspec/rspec-rails/pull/2774 which tries to keep within the spirit of implementation Jon describes and provides a compromise solution that should work for pretty much everyone. Details in the PR. Feedback very welcome! If the implementation isn't controversial, I can imagine that the test code might be...

Should I be updating the topmost 'development' section of `Changelog.md` in PRs too, or is that done by whoever merges stuff?

**Comment by pirj** [comment]
[This](https://github.com/rspec/rspec-rails/issues/2713#issuecomment-2041112990) seems related, but the [related PR](https://github.com/rspec/rspec-rails/pull/2753) stalled.
